### PR TITLE
Mesh Divisions system to split the mesh in arbitrary ways

### PR DESCRIPTION
**Pull request by GiudGiud**
refs https://github.com/idaholab/moose/issues/25901

I'm delivering this with quite a few items to hope to see some adoption. I think it worked for Times & Positions to do that.
Realistically while this makes for a long review it also makes it efficient because the logic is similar everywhere

I am not including the hex lattice (which is the one I need) because there's a need for a utility and we should do this right once and for all. I can make do with nearest-positions for the transfer most likely.

### CODE REVIEW

**Comment by GiudGiud** [comment]
Threaded failure is real and I'm thinking about it but otherwise this can be looked at

**Comment by GiudGiud** [comment]
I can add more testing (with a new potentially useful object) once https://github.com/idaholab/moose/pull/25872 is in

File: framework/src/meshdivisions/NearestPositionsDivision.C [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+registerMooseObject("MooseApp", NearestPositionsDivision);
+
+InputParameters
+NearestPositionsDivision::validParams()
+{
+  InputParameters params = MeshDivision::validParams();
+  params.addClassDescription("Divide the mesh by ");
+  params.addRequiredParam<PositionsName>(
+      "positions",
+      "The name of the Positions object to form the nearest-neighbor partition/division with");
```
Comment (lindsayad) [inline framework/src/meshdivisions/NearestPositionsDivision.C:26]:
```suggestion
      "The name of the Positions object to form the nearest-neighbor division with");
```

File: framework/src/meshdivisions/NearestPositionsDivision.C [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+void
+NearestPositionsDivision::initialize()
+{
+  setNumDivisions(_nearest_positions_obj->getNumPositions());
+}
+
+unsigned int
+NearestPositionsDivision::divisionIndex(const Elem & elem) const
+{
+  bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
```
Comment (lindsayad) [inline framework/src/meshdivisions/NearestPositionsDivision.C:46]:
```suggestion
  const bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
```

File: framework/src/meshdivisions/NearestPositionsDivision.C [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  return _nearest_positions_obj->getNearestPositionIndex(elem.vertex_average(), initial);
+}
+
+unsigned int
+NearestPositionsDivision::divisionIndex(const Point & pt) const
+{
+  bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
```
Comment (lindsayad) [inline framework/src/meshdivisions/NearestPositionsDivision.C:53]:
```suggestion
  const bool initial = _fe_problem->getCurrentExecuteOnFlag() == EXEC_INITIAL;
```

File: framework/src/meshdivisions/NestedDivision.C [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  auto tot_divs = 1;
+  _num_divs.resize(_divisions.size());
+
+  for (const auto i : index_range(_divisions))
+  {
+    _num_divs[i] = _divisions[i]->getNumDivisions();
+    tot_divs *= _num_divs[i];
+    if (_num_divs[i] == 0)
+      mooseError("Nested division",
+                 getParam<std::vector<MeshDivisionName>>("divisions")[i],
```
Comment (lindsayad) [inline framework/src/meshdivisions/NestedDivision.C:46]:
```suggestion
                 _divisions[i]->name(),
```

File: framework/src/meshdivisions/NestedDivision.C [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                 "now has zero bins");
```
Comment (lindsayad) [inline framework/src/meshdivisions/NestedDivision.C:47]:
```suggestion
                 "' has zero bins");
```

File: framework/src/meshdivisions/NestedDivision.C [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+{
+  auto tot_divs = 1;
+  _num_divs.resize(_divisions.size());
+
+  for (const auto i : index_range(_divisions))
+  {
+    _num_divs[i] = _divisions[i]->getNumDivisions();
+    tot_divs *= _num_divs[i];
+    if (_num_divs[i] == 0)
+      mooseError("Nested division",
```
Comment (lindsayad) [inline framework/src/meshdivisions/NestedDivision.C:45]:
```suggestion
      mooseError("Nested division '",
```

File: framework/src/meshdivisions/NestedDivision.C [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  }
+  setNumDivisions(tot_divs);
+}
+
+unsigned int
+NestedDivision::divisionIndex(const Elem & elem) const
+{
+  unsigned int index = 0;
+  unsigned int running_product = 1;
+  auto N_divs = _divisions.size();
```
Comment (lindsayad) [inline framework/src/meshdivisions/NestedDivision.C:57]:
```suggestion
  const auto N_divs = _divisions.size();
```

File: framework/src/meshdivisions/NestedDivision.C [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  }
+  return index;
+}
+
+unsigned int
+NestedDivision::divisionIndex(const Point & pt) const
+{
+  unsigned int index = 0;
+  unsigned int running_product = 1;
+  auto N_divs = _divisions.size();
```
Comment (lindsayad) [inline framework/src/meshdivisions/NestedDivision.C:71]:
```suggestion
  const auto N_divs = _divisions.size();
```

File: framework/src/meshdivisions/SphericalGridDivision.C [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+  if (!_outside_grid_counts_as_border)
+  {
+    if (MooseUtils::absoluteFuzzyLessThan(pc(0), _min_r) ||
+        MooseUtils::absoluteFuzzyGreaterThan(pc(0), _max_r))
+      return MooseMeshDivision::INVALID_DIVISION_INDEX;
+  }
+
+  const auto not_found = MooseMeshDivision::INVALID_DIVISION_INDEX;
+  unsigned int ir = not_found;
```
Comment (lindsayad) [inline framework/src/meshdivisions/SphericalGridDivision.C:90]:
```suggestion
  auto ir = not_found;
```

File: framework/src/postprocessors/NumMeshDivisions.C [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+// MOOSE includes
+#include "NumMeshDivisions.h"
+#include "MeshDivision.h"
```
Comment (lindsayad) [inline framework/src/postprocessors/NumMeshDivisions.C:12]:
```suggestion
#include "MeshDivision.h"
#include "FEProblemBase.h"
```

File: test/tests/meshdivisions/tests [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+      csvdiff = 'spherical_division_out_div_out_0001.csv'
+      detail = 'a regular spherical radial grid,'
+      # Sampling is not ordered the same way on distributed meshes
+      mesh_mode = REPLICATED
+    []
+    [nested]
+      type = CSVDiff
+      input = 'nested_division.i'
+      csvdiff = 'nested_division_out_div_out_0001.csv'
+      detail = 'a nesting of mesh divisions.'
```
Comment (lindsayad) [inline test/tests/meshdivisions/tests:64]:
```suggestion
      detail = 'and a nesting of mesh divisions.'
```

File: unit/src/MooseServerTest.C [hunk 12]
```diff
@@ -933,7 +933,7 @@ TEST_F(MooseServerTest, CompletionDocumentRootLevel)
 
   EXPECT_EQ(request_id, response_id);
 
-  EXPECT_EQ(46u, completions_array.size());
+  EXPECT_EQ(47u, completions_array.size());
```
Comment (lindsayad) [inline unit/src/MooseServerTest.C:936]: I agree with your issue that this is unfortunate

File: framework/doc/content/source/meshdivisions/NearestPositionsDivision.md [hunk 13]
```diff
@@ -0,0 +1,18 @@
+# NearestPositionsDivision
+
+!syntax description /MeshDivisions/NearestPositionsDivision
+
+The number of divisions is simply equal to the number of positions in the [Positions](syntax/Positions/index.md)
+object specified in the [!param](/MeshDivisions/NearestPositionsDivision/positions) parameter.
+
+!alert note
+For points laying within the standard tolerance of an internal boundary of the bins,
```
Comment (lindsayad) [inline framework/doc/content/source/meshdivisions/NearestPositionsDivision.md:9]:
```suggestion
For points lying within the standard tolerance of an internal boundary of the bins,
```

File: framework/doc/content/source/meshdivisions/SphericalGridDivision.md [hunk 14]
```diff
@@ -0,0 +1,25 @@
+# SphericalGridDivision
+
+!syntax description /MeshDivisions/SphericalGridDivision
+
+The number of mesh divisions/regions is the product of the number of bins in each coordinate, which currently only
+includes the radial direction.
+
+Points that lay outside the spherical grid may be assigned to the closest grid outer bins (on the inner or outer shell)
```
Comment (lindsayad) [inline framework/doc/content/source/meshdivisions/SphericalGridDivision.md:8]:
```suggestion
Points that lie outside the spherical grid may be assigned to the closest grid outer bins (on the inner or outer shell)
```

File: framework/doc/content/source/meshdivisions/SphericalGridDivision.md [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+using the [!param](/MeshDivisions/SphericalGridDivision/assign_domain_outside_grid_to_border)
+parameter.
+
+!alert note
+We have not implemented neither binning in the azimuthal and toroidal coordinates nor restrictions in those angular coordinates.
```
Comment (lindsayad) [inline framework/doc/content/source/meshdivisions/SphericalGridDivision.md:13]:
```suggestion
We have not implemented binning in the azimuthal nor toroidal coordinates nor restrictions in those angular coordinates.
```

File: framework/doc/content/source/meshdivisions/SphericalGridDivision.md [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+This is a desirable extension of this object.
+
+!alert note
+For points laying within the standard tolerance of an internal boundary of the spherical grid, this object
```
Comment (lindsayad) [inline framework/doc/content/source/meshdivisions/SphericalGridDivision.md:17]:
```suggestion
For points lying within the standard tolerance of an internal boundary of the spherical grid, this object
```

**Review by lindsayad — CHANGES_REQUESTED** [review]
looks great!

File: framework/include/meshdivisions/FunctorBinnedValuesDivision.h [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+  unsigned int divisionIndex(const Elem & elem) const override;
+
+  /**
+   * @brief Get the bin for that functor value
+   * @param value functor value to bin
+   * @param pt point where the functor is being evaluated, useful for warnings
+   */
+  unsigned int getBinIndex(Real value, const Point & pt) const;
+
+protected:
```
Comment (GiudGiud) [inline framework/include/meshdivisions/FunctorBinnedValuesDivision.h:36]:
```suggestion
```

File: framework/include/meshdivisions/FunctorBinnedValuesDivision.h [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+  void initialize() override;
+  unsigned int divisionIndex(const Point & pt) const override;
+  unsigned int divisionIndex(const Elem & elem) const override;
+
+  /**
+   * @brief Get the bin for that functor value
+   * @param value functor value to bin
+   * @param pt point where the functor is being evaluated, useful for warnings
+   */
+  unsigned int getBinIndex(Real value, const Point & pt) const;
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/include/meshdivisions/FunctorBinnedValuesDivision.h:34]: no. moving it

File: framework/include/meshdivisions/FunctorBinnedValuesDivision.h [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+{
+public:
+  static InputParameters validParams();
+
+  FunctorBinnedValuesDivision(const InputParameters & parameters);
+
+  void initialize() override;
+  unsigned int divisionIndex(const Point & pt) const override;
+  unsigned int divisionIndex(const Elem & elem) const override;
+
```
Comment (GiudGiud) [inline framework/include/meshdivisions/FunctorBinnedValuesDivision.h:28]:
```suggestion
protected:

```

File: framework/include/meshdivisions/MeshDivision.h [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MooseObject.h"
+#include "SetupInterface.h"
+#include "Restartable.h"
+#include "MeshChangedInterface.h"
+
+// libMesh
```
Comment (GiudGiud) [inline framework/include/meshdivisions/MeshDivision.h:17]:
```suggestion
```

File: framework/src/auxkernels/MeshDivisionAux.C [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+MeshDivisionAux::MeshDivisionAux(const InputParameters & parameters)
+  : AuxKernel(parameters),
+    _mesh_division(_c_fe_problem.getMeshDivision(getParam<MeshDivisionName>("mesh_division"), _tid))
+{
+  // Check family types for reasonable choices
+  if (isNodal() && _var.feType().order != FIRST)
+    paramError("Variable order " + Moose::stringify(_var.feType().order) +
+               " unsupported for nodal variable");
```
Comment (GiudGiud) [inline framework/src/auxkernels/MeshDivisionAux.C:33]:
```suggestion
```

File: framework/src/auxkernels/MeshDivisionAux.C [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+  else if (_var.feType().order != CONSTANT)
```
Comment (GiudGiud) [inline framework/src/auxkernels/MeshDivisionAux.C:34]:
```suggestion
  if (!_var.isNodal() && _var.feType().order != CONSTANT)
```

File: framework/src/meshdivisions/NearestPositionsDivision.C [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+#include "libmesh/mesh_base.h"
+#include "libmesh/elem.h"
+
+registerMooseObject("MooseApp", NearestPositionsDivision);
+
+InputParameters
+NearestPositionsDivision::validParams()
+{
+  InputParameters params = MeshDivision::validParams();
+  params.addClassDescription("Divide the mesh by ");
```
Comment (GiudGiud) [inline framework/src/meshdivisions/NearestPositionsDivision.C:23]:
```suggestion
  params.addClassDescription("Divide the mesh using a nearest-point / voronoi algorithm, with the points coming from a Positions object");
```

**Review by lindsayad — APPROVED** [review]

**Pull request merged**